modules/core/src/main/java/org/apache/ignite/internal/processors/cache/consistentcut/ConsistentCut.java

```diff
@@ -0,0 +1,260 @@
+
+                if (!(tx.state() == UNKNOWN
+                    || tx.state() == MARKED_ROLLBACK
+                    || tx.state() == ROLLED_BACK
+                    || tx.state() == COMMITTED)) {
+                    U.warn(log, "Transaction is in unexepcted state: " + tx.state() +
+                        ". Cut might be inconsistent. Transaction: " + tx);
+
+                    return false;
+                }
+
+                if (tx.state() == UNKNOWN) {
+                    U.warn(log, "Transaction is in UNKNOWN state. Cut might be inconsistent. Transaction: " + tx);
+
+                    return false;
+                }
+
+                // ROLLED_BACK transactions don't change data then don't care.
+                if (tx.state() == ROLLED_BACK || tx.state() == MARKED_ROLLBACK)
+                    return true;
+
+                if (tx.finalizationStatus() == RECOVERY_FINISH) {
+                    U.warn(log, "Transaction committed after recovery process. Cut might be inconsistent. Transaction: " + tx);
+
+                    return false;
+                }
+
+                if (id.equals(tx.cutId()))
+                    afterCut.add(tx.nearXidVersion());
+                else
+                    beforeCut.add(tx.nearXidVersion());
+
+                return true;
+            });
+
+            checkFut.add(txCheckFut);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected boolean onDone(@Nullable WALPointer res, @Nullable Throwable err, boolean cancel) {
+        removedActive = null;
+
+        return super.onDone(res, err, cancel);
+    }
+
+    /**
+     * Logs Consistent Cut Record to WAL.
+     *
+     * @param record Record to write to WAL.
+     * @param finalRec {@code false} for {@link ConsistentCutStartRecord}, {@code true} for {@link ConsistentCutFinishRecord}.
+     * @return Pointer to the record in WAL, or {@code null} if WAL is disabled.
+     */
+    private @Nullable WALPointer walLog(WALRecord record, boolean finalRec) throws IgniteCheckedException {
+        if (cctx.wal() != null) {
+            if (log.isDebugEnabled())
+                log.debug("Writing Consistent Cut WAL record: " + record);
+
+            if (finalRec) {
+                cctx.database().checkpointReadLock();
+
+                try {
+                    return cctx.wal().log(record, RolloverType.CURRENT_SEGMENT);
+                }
+                finally {
+                    cctx.database().checkpointReadUnlock();
+                }
+            }
+            else
+                return cctx.wal().log(record);
+        }
+
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString() {
+        List<IgniteUuid> before = null;
+        List<IgniteUuid> after = null;
+
+        // Write IgniteUuid because it's more convenient for debug purposes than GridCacheVersion.
+        if (beforeCut != null) {
+            before = beforeCut.stream()
```

Comment:
It we need to do some transformation on `before`, `after` sets before printing them out, let's make some utility method for this purpose.